Four pipe beam air terminal

EnergyArchmage opened the pull request:
This new feature for chilled and heated beam is ready to go in I believe.  Code, unit test and docs are all completed. 

Please review and merge into develop. 

Includes a large amount of clear_state stuff for unit test fixture.

EnergyArchmage committed:
first draft

EnergyArchmage committed:
add unit tests, major work on test fixtures

adding clear_state to many files, added unit test for four pipe beam.
causes problem to

EnergyArchmage committed:
fix missed PlantLocation change, fix OutputProcessor,

more clear_state routines. fixes.

EnergyArchmage committed:
merge in develop

EnergyArchmage committed:
clean up merge of develop

fixed some collisions.  update GlobalNames for new style

EnergyArchmage committed:
operator assignment fix

fix from Stuart. complies on windows now.

EnergyArchmage committed:
fix build warnings, expand unit test

EnergyArchmage committed:
test file, I/O ref, Eng Ref

add integration test file.  draft I/O ref changes, and eng ref changes.
separate out adjustments for design heat and cool flow rates.  use sea
level rho for normalized, rated flow rate data.

EnergyArchmage committed:
Merge branch 'develop' into FourPipeBeamAirTerminal

EnergyArchmage committed:
add to tips and tricks

EnergyArchmage committed:
Expand test fixture

attempting to get duplicate unit test to pass, major additional flush
out of test fixture clear_state.  still not working though

EnergyArchmage committed:
additional fixture fix

finally got unit test duplicate to complete as expected.

EnergyArchmage committed:
Merge branch 'develop' into FourPipeBeamAirTerminal

EnergyArchmage committed:
more eng ref changes

EnergyArchmage committed:
protect not present beam; A to Z order in fixture

add additional protection when beam cooling or heat is not present.
reorder includes and clear state to be A to Z in fixture

EnergyArchmage committed:
Merge branch 'develop' into FourPipeBeamAirTerminal

EnergyArchmage committed:
clean up I/O ref and unit test

ready for pull request now

mbadams5 commented on src/EnergyPlus/AirTerminalUnit.hh:
```diff
@@ -0,0 +1,118 @@
+	// for unit tests
+	virtual
+	void
+	clear_state() = 0;
```
@EnergyArchmage I don't think there should be a clear_state() within classes. The clear_state() was a (hacky) way to fix unit testing with so many namespace variables. You won't need to clear any state within an object, and if you do, it should be in the destructor (should be rare we need to do this).

mbadams5 commented on src/EnergyPlus/HVACFourPipeBeam.cc:
```diff
@@ -0,0 +1,1302 @@
+	void
+	HVACFourPipeBeam::clear_state()
+	{
+		FourPipeBeams.deallocate();
```
@EnergyArchmage This is why we shouldn't have a clear_state() in classes...

Are all objects calling clear_state() or just one? What happens if you call clear_state on one object and then go to a different object and do something else? Really this clear_state() should be a function within the namespace like the other clear_states, not within the class.

mbadams5 commented on src/EnergyPlus/HVACFourPipeBeam.hh:
```diff
@@ -0,0 +1,280 @@
+public: // Creation
+
+	// Default Constructor
+	HVACFourPipeBeam() :
```
@EnergyArchmage If you have a static Factory method to create this object, why is the constructor public? Should it be private (or protected)? It should work to be private but I would have to double check to see if it compiles for sure.

EnergyArchmage replied:
The constructor is called from ZoneAirLoopEquipmentManager.cc, so I figure it has to be public.

mbadams5 replied:
@EnergyArchmage You are calling the factory method in ZoneAirLoopEquipmentManager.cc, not the class constructor. I guess the question is whether we want the constructors to be public or private (if they truly can be private)? @Myoldmopar @kbenne thoughts?

Myoldmopar replied:
I think if it is intended to only be used from within the factory, it would make more sense to keep it private.  The factory is needed to ensure you don't double instantiate things, so no reason to allow an external call to the constructor directly.  I'd recommend private at this point.  But I think (maybe?) there are further comments on this below?

mbadams5 commented on src/EnergyPlus/HVACFourPipeBeam.hh:
```diff
@@ -0,0 +1,280 @@
+	void
+	report() ;
+
+public: // Data
```
@EnergyArchmage Does this data need to be public? I haven't looked fully to see whether it does or not. Is it only used within this class or does it need to be accessed and used outside this class as well?

EnergyArchmage replied:
It can be private, making that change now.

mbadams5 commented on src/EnergyPlus/HVACFourPipeBeam.cc:
```diff
@@ -0,0 +1,1302 @@
+
+		if ( ( originalTermUnitSizeMaxVDot > 0.0 ) && ( originalTermUnitSizeMaxVDot != this->vDotDesignPrimAir ) && ( CurZoneEqNum > 0 ) ) {
+			// perturb system size to handle chnage in system size calculated without knowing about 4 pipe beam 
+			DataSizing::FinalSysSizing( DataZoneEquipment::ZoneEquipConfig(CurZoneEqNum).AirLoopNum ).DesMainVolFlow 
```
@EnergyArchmage Here you can just do...

``` CPP
DataSizing::FinalSysSizing( DataZoneEquipment::ZoneEquipConfig(CurZoneEqNum).AirLoopNum ).DesMainVolFlow += 
this->vDotDesignPrimAir - originalTermUnitSizeMaxVDot;
```

instead which is less code, but also better since it avoids another lookup.

You can do this for the following few sizing calls.

mbadams5 commented on src/EnergyPlus/HVACFourPipeBeam.cc:
```diff
@@ -0,0 +1,1302 @@
+
+		NonAirSysOutput = 0.0; // initialize
+
+		if ( this->mDotSystemAir == 0.0 || ( ! this->airAvailable  && ! this->coolingAvailable && ! this->heatingAvailable ) ) { //unit is off
```
@EnergyArchmage This isn't specific to you but is something systemic to EnergyPlus as a whole. Doing this comparison, `mDotSystemAir == 0.0` can cause subtle issues. Floating point comparisons are finicky and hard to do correctly. We need an approach for EnergyPlus as a whole. 

This issue with that comparison is it will evaluate to false if `mDotSystemAir = 0.0000000000000001` which it could due to floating point rounding error. 

@Myoldmopar fyi

mjwitte replied:
@mbadams5 We do have [globals](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/DataHVACGlobals.cc#L31-L37) for these types of tolerances.  Although I'm not sure 0.001 is small enough and 1e-30 may be too small.

mbadams5 commented on src/EnergyPlus/HVACFourPipeBeam.cc:
```diff
@@ -0,0 +1,1302 @@
+			return;
+		}
+
+		return;
```
@EnergyArchmage You don't really need this return here, but it doesn't hurt anything.

mbadams5 commented on tst/EnergyPlus/unit/Fixtures/HVACFixture.hh:
```diff
@@ -6,6 +6,24 @@
 
 // EnergyPlus Headers
 #include "EnergyPlusFixture.hh"
+#include <EnergyPlus/CurveManager.hh>
```
@EnergyArchmage Since you moved all the clear_state() methods to the EnergyPlusFixture TearDown(), you can (and should) get rid of these includes here. You might need to put includes into specific unit test files again but that will show up when you go to compile.

EnergyArchmage replied:
Note that clear_state tear downs are not the only reason for many of the includes; sometimes the unit tests are accessing routines and data.  I will make this change but it seems to me to make HVACFixture less useful because then the includes need to be repeated in different unit tests.

mbadams5 commented on tst/EnergyPlus/unit/HVACFourPipeBeam.unit.cc:
```diff
@@ -0,0 +1,1705 @@
+
+		"  NodeList,",
+		"    HeatSys1 Loop Setpoint Node List,  !- Name",
+		"    HeatSys1 Supply Outlet Node;  !- Node 1 Name",
```
@EnergyArchmage I know doubles get automatically converted to floats, but you should probably use EXPECT_DOUBLE_EQ here instead since the variable is a double (Real64).

mbadams5 replied:
@EnergyArchmage Was this required for your unit test to work properly? This could be something that I could possibly add into the process_idf() function.

EnergyArchmage replied:
I can't tell what this comment is referring to.  If it is the call to SimulationManager::PostIPProcessing, then yes I think it should probably go into the process_idf function.

mbadams5 replied:
Yes it was. I don't know how or why it moved so much...

Myoldmopar replied:
I wonder if you had an out-of-date 'files changed' tab open and it got confused...

mbadams5 commented on src/EnergyPlus/WeatherManager.cc:
```diff
@@ -287,6 +296,241 @@ namespace WeatherManager {
+		WeatherFileTimeZone = 0.0 ;
+		WeatherFileElevation = 0.0 ;
+		WeatherFileUnitNumber = 0 ; // File unit number for the weather file
+		for ( int i = 1; i <= 12; i++ )
```
@EnergyArchmage Instead of a for loop, you can do this...

``` cpp
GroundTemps = Array1D< Real64 >( 12, 18.0 );
GroundTempsFC = Array1D< Real64 >( 12, 0.0 );
```

and so on for all the variables you have in the two for loops.

mbadams5 commented on src/EnergyPlus/HVACFourPipeBeam.cc:
```diff
@@ -0,0 +1,1302 @@
+
+namespace FourPipeBeam {
+
+	Array1D< std::shared_ptr< HVACFourPipeBeam > > FourPipeBeams; // dimension to number of machines
```
@EnergyArchmage Do you intend for these HVACFourPipeBeam objects to be used and referenced by other objects? Basically do these need to be std::shared_ptr or can they be std::unique_ptr? They very well could require the use of std::shared_ptr. 

For example, will an air terminal or any other object hold a `std::shared_ptr< HVACFourPipeBeam >` object within their class?

@Myoldmopar fyi

EnergyArchmage replied:
I think the answer is yes, they are used in a structure defined in DataDefineEquip.hh

```
       struct ZoneAirEquip
              std::shared_ptr< AirTerminalUnit > airTerminalPtr;
```

This is how we can get to OO method calls from the rest of the program.  I took this pattern from Edwin’s branch PullLogicFromPlantComponents.  

But note the type there is the base case AirTerminalUnit, not these HVACFourPipeBeam objects, so that in the future other types of air terminal units can use the same thing.

mbadams5 replied:
I agree with you. I missed that apparently.

DeadParrot replied:
@EnergyArchmage _et al_, we should set up some standards for pointer use before the anti-pattern of using `shared_ptr` everywhere becomes widespread. There is considerable overhead to `shared_ptr` such that it should only be used in contexts where shared ownership is natural.

Myoldmopar replied:
@DeadParrot The main reason I chose to use shared_ptr in the plant restructuring branch I have is that a single component can live on multiple plant loops, but each loop uses the same entry point into the calculation, so it was a natural fit.  There was a positive response from many folks about using this, but I think we'd all be happy to have a little more discussion before the pattern becomes mainstream.

One comment on this though.  I think we have a little bit of a chicken/egg with regard to trying out these concepts and figuring out what is the best design.  I put out the shared_ptr approach in my branch, and it works great.  I like letting @EnergyArchmage try it here because he will be vocal on whether it seems completely against the grain.  If we find it doesn't work in places, or there is a noticeable performance hit that outweighs the benefit of using them, then we aren't going to be very far down the road in terms of pulling back.  And at that point we would have more data points to determine what would be _better_.

All that said, let's have a chat maybe next week?  Or would it be better suited for immediately after the release?

Myoldmopar replied:
@EnergyArchmage with my description [here](https://github.com/NREL/EnergyPlus/pull/5129#discussion_r37910514) of how I came to using shared_ptr because multiple plant loops can _own_ a single plant component instance, can you comment on whether one of these terminals will be _owned_ by multiple managers in the code?  Or will it always be called from a single spot.  

Note this is a different discussion than the polymorphism you described between `AirTerminalUnit` and `HVACFourPipeBeam`.  I'm happy to chat further, just let me know!

DeadParrot replied:
@Myoldmopar An entity that doesn't have a natural owner can indeed be a natural use of `shared_ptr`. While we are still having "master" containers we don't need shared ownership: an owning (unique or raw) pointer in the master container and exposing only via reference in the public interface works fine. As we move away from master containers we may need some shared ownership but using them everywhere is a common mistake: they create problems beyond performance and, unfortunately, they aren't a magic wand to make memory management easy. I encourage reading up on the drawbacks of shared pointers followed by someone writing up a proposed guideline on memory/ownership and a discussion. Until then I suggest that any use of `shared_ptr` must be justified and alternatives evaluated. The problems with "works great so far" are that we aren't assessing performance, code will be used as a model by the next developer, and shared ownership pointer issues snowball as they spread out in large projects: this really needs to be thought through with care before we go too far. Using shared pointers widely and waiting until we find we've created a problem is a mistake.

EnergyArchmage replied:
Here there is only 1 master, the AirLoopHVAC, for the foreseeable future.  So I guess there is no special need for a shared pointer.  

I suppose the alternative is obvious to you all, but I can only guess it is to only instantiate the objects in struct ZoneAirEquip and have no such declaration in namespace HVACFourPipeBeam.  

I am happy to help set a good pattern here, but not keen for this project to be held hostage to the perfect.

mbadams5 commented on src/EnergyPlus/GlobalNames.cc:
```diff
@@ -334,6 +358,41 @@ namespace GlobalNames {
+				aDUNames.allocate( curMaxAirDistUnits );
+			} else if ( numAirDistUnits == curMaxAirDistUnits ) {
+				curMaxAirDistUnits += 4;
+				aDUNames.redimension( curMaxAirDistUnits );
```
@EnergyArchmage For new code, i'd suggest moving to a `push_back()` approach instead of `redimension()` since @DeadParrot added that functionality earlier this summer.

``` cpp
ComponentNameData aDUdata( MakeUPPERCase( TypeToVerify ), NameToVerify );

// This has a member constructor that works but if not then you do this...
// aDUdata.CompType = MakeUPPERCase( TypeToVerify );
// aDUdata.CompName = NameToVerify;

aDUNames.push_back( aDUdata );
// could use aDUNames.emplace_back( MakeUPPERCase( TypeToVerify ), NameToVerify );
// this would avoid both lines above and is better.

++numAirDistUnits;
```

This code snippet would replace everything in the `else` statement. This is a MUCH better approach than the old Fortran type approach. It is better for performance (speed and memory) and readability.

@DeadParrot As an aside, i've had some trouble with emplace_back() when i've played around with it locally. Sometimes it doesn't work, but I can see if I can find a concrete example to send to you.

DeadParrot replied:
@mbadams5 If you already have an item of the array's `value_type` I suggest just using `push_back` since there is a "move" overload of it. The `emplace_back` is currently a quick placeholder and I hope to refine that and add some more vector-like API in the next couple of months. But if you have an example that should work but doesn't please send it to me. I do have a 2-iterator constructor waiting in the wings, though, which should be useful as we start using more container types.

I second the preference for `push_back` in new code (and migrating the `redimension` calls to it gradually) but a couple of things to note:
- The `curMaxAirDistUnits` then grows by 1 every time
- You can either `push_back` a default-contstructed element and then set the `CompType` and `CompName` or create such an element first and push it back. (If you could create it as a temporary inside the `push_back` that would let C++ do a move with potential performance benefits, but in practice most of our structs are still buckets of PODs so moving isn't going to make a difference yet.)
- Moving from `redimension` to `push_back` can "break" tests by changing the size of arrays that are reported in the audit files: that is OK but should be noted in the PR

mbadams5 commented on src/EnergyPlus/AirTerminalUnit.hh:
```diff
@@ -0,0 +1,118 @@
+
+
+
+public: // Data
```
@EnergyArchmage Does the follow data need to be public or can it be protected? Does anything outside of the derived classes need to access this data?

EnergyArchmage replied:
The unit tests need to access the data and will not compile with protected.

mbadams5 replied:
@EnergyArchmage I guess I should be more specific. Does anything outside of the derived classes need to access that data within EnergyPlus simulation, not unit tests?

I personally think making things public to make unit testing easier isn't the best design. The member functions and variables that are public should be thought of as a contract with users of your class. These are the public API of your class and should only be changed when the cost of not changing them is greater than the cost of changing them. Anything else is a private implementation detail and should not be exposed. This allows you to change implementations without affecting the public API. 

This starts to get into unit testing paradigms but unit tests should usually only test public APIs. If something is private/protected and needs to be unit tested for valid reasons (very complex, critical but not user facing function, etc) then you can `friend` a Fixture. I've had to do this for SQLiteProcedures and my process_idd() function. But other people might have different opinions.

All that said, if these member variables truly need to be public then all I said does not apply in this case.

EnergyArchmage replied:
I think protected would work fine in the base class as far as the main program is concerned; the derived class won't seem to comple with private data in the base class.

I did already change all the data to be private in the derived class.   

I could probably remove and/or change checks in the unit test to only check stuff outside the class.  But hiding everything seems to limit what unit testing can do, sans friend that is.  I can't really judge what to do here without some specific guidance.

mbadams5 replied:
@EnergyArchmage Yeah in the base class it would have to be protected if a derived class is using that variable. Basically, private makes it so ONLY that class has access while protected allows that classes and any derived classes access to that data. 

Lets try to pull some more people and their thoughts in on this... @Myoldmopar @kbenne

EnergyArchmage replied:
Okay, I just pushed up changes to use protected in the base class and then friend class HVACFixture and it seems to work.

mbadams5 commented:
@EnergyArchmage Overall this looks really good. I am curious to see what others think about the OOP design. We need to come to a good design pattern. I will review the OOP design later when I have more time to think about it. At first glance, it seems good, but the devil is always in the details.

Also, a huge :clap: :+1: on adding all the clear_state() and the unit tests. Hopefully the recent changes helped to make the unit tests easier and quicker to write (after adding all the clear_state()).

EnergyArchmage committed:
use push_back in new GlobalNames code

address comments

EnergyArchmage committed:
remove for loops in clear state WeatherManager

address comment from reviews

EnergyArchmage committed:
remove clear_state from new classes

address review comments

EnergyArchmage committed:
private data, clean up eng ref, clean up output variables

make data and some method private in HVACFourPipeBeam object, fix
inadvertent degree symbol changes in eng. ref.  clean up output:variable
requests in integration test file.

EnergyArchmage committed:
use VerySmallMassFlow instead of 0.0

revise comparisons to 0.0 to use DataHVACGlobals::VerySmallMassFlow

EnergyArchmage committed:
Merge branch 'develop' into FourPipeBeamAirTerminal

EnergyArchmage committed:
change HVACFixture includes, use +=

addressing comments on #5129

EnergyArchmage commented:
I think I have addressed all the comments.  assigning to edwin.

mbadams5 commented on tst/EnergyPlus/unit/HVACFourPipeBeam.unit.cc:
```diff
@@ -0,0 +1,1716 @@
+		"    HeatSys1 Loop Setpoint Manager,  !- Name",
+		"    Temperature,             !- Control Variable",
+		"    HeatSys1 Loop Temp Sch,    !- Schedule Name",
+		"    HeatSys1 Loop Setpoint Node List;  !
```
Copy from above since it for some reason moved...

@EnergyArchmage I know doubles get automatically converted to floats, but you should probably use EXPECT_DOUBLE_EQ here instead since the variable is a double (Real64).

EnergyArchmage committed:
EXPECT_DOUBLE_EQ, friend and protected base class

changed unit test to use EXPECT_DOUBLE_EQ, instead of EXPECT_FLOAT_EQ.
made base class data protected and added friend to HVACFixture.
clear_state in VS coil module.

EnergyArchmage committed:
EXPECT_NEAR, protected constructor, protected data

change some EXPECT_DOUBLE_EQ to EXPECT_NEAR.  change public to protected
in two places (correct accident from last commit)

EnergyArchmage committed:
Merge branch 'develop' into FourPipeBeamAirTerminal

EnergyArchmage committed:
change beam constructor to private, extend clear_state for VRFTest

fixed missed EXPECT_DOUBLE_EQ.  add missing end line comments.  change
HVACFourPipeBeam object constructor from protected to private.
add negative comment about continued use of shared_ptr. extend
clear_state changes to get VRFTest working

Myoldmopar committed:
Merged develop in; attempted resolution of conflicts

Myoldmopar committed:
Update compiler warnings, clean up duplicated clear_state()s

Myoldmopar commented:
@EnergyArchmage I cleaned up the two compiler warnings, and I did a merge from develop, resolving a few conflicts that seemed straightforward.  I got it to build fine, and the unit tests run fine, but the new example file is failing for me as it appears it was before.  The program crashes with divide by zero [on this line](https://github.com/NREL/EnergyPlus/blob/FourPipeBeamAirTerminal/src/EnergyPlus/HVACFourPipeBeam.cc#L1137) with `this->mDotHW` equal to zero.  Probably just need to add protection before this calculation?

EnergyArchmage committed:
protect div by zero in four pipe beam

EnergyArchmage committed:
Merge branch 'develop' into FourPipeBeamAirTerminal

resolved Conflicts in following files,  moving clear state calls to
EnergyPlusFixture per earlier decision by EL and MA.
	src/EnergyPlus/GlobalNames.cc
	src/EnergyPlus/GlobalNames.hh
	tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc
	tst/EnergyPlus/unit/Fixtures/HVACFixture.hh

EnergyArchmage committed:
Merge remote-tracking branch 'origin/FourPipeBeamAirTerminal' into FourPipeBeamAirTerminal

try again to resolve Conflicts:
	src/EnergyPlus/Fans.cc
	src/EnergyPlus/GlobalNames.cc
	src/EnergyPlus/GlobalNames.hh
	tst/EnergyPlus/unit/Fixtures/EnergyPlusFixture.cc
	tst/EnergyPlus/unit/Fixtures/HVACFixture.hh

EnergyArchmage commented:
okay I add some div by zero corrections and did another round of merging in develop and resolving clear_state collisions.  duplicate efforts on clear_state going on.

Myoldmopar commented:
@EnergyArchmage @mjwitte Got this merged up with develop locally; just a couple pretty straightforward conflicts.  I'm building to make sure everything is OK and I'll re-push.  The CI tests were looking good above with a3c4404 though, so this might be ready to slide in??

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39101,6 +39101,170 @@ AirTerminal:SingleDuct:ConstantVolume:FourPipeInduction,
+       \minimum> 0.0
+       \autosizable
+       \default autosize
+  N5 , \field Rated Primary Air Flow Rate per Meter
```
@EnergyArchmage There are several (four?) fields that have a similar "... per Meter" name.  Should probably be " . . . per Beam Length".

mjwitte commented:
@Myoldmopar @EnergyArchmage That's my only comment.

Myoldmopar committed:
Merge develop; resolve conflict

Myoldmopar commented:
@EnergyArchmage Alright, I got the merge to build.  If CI comes up happy, I will move forward assuming the coding is complete.  If you can adjust the field names that @mjwitte mentioned above in the IDD and docs, we'll be good to go in.

EnergyArchmage committed:
change field names to "per Beam Length"

changed Meter to Beam Length in five places, idd and I/O ref

mjwitte commented:
@EnergyArchmage There are some build warnings and a unit test failure.  Are you able to look these this afternoon?

EnergyArchmage committed:
clean up unused vars, fix DOAS size unit test error.

attempting to fix unused variable warnings.   changes to get unit test
DOASEffectOnZoneSizing.SizeZoneEquipment to work.

EnergyArchmage commented:
Sorry I was traveling and could not get to it until today (finally made it back to CO!). I just got to those unused vars and DOAS size unit test fix and committed.

EnergyArchmage committed:
Merge branch 'develop' into FourPipeBeamAirTerminal

mjwitte commented:
@EnergyArchmage No problem.  We had plenty of other stuff to work on.  Thanks for the updates.
